tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc

```diff
@@ -0,0 +1,447 @@
+    // Do not fuse computation with too large shapes or too many instructions,
+    // because it may not be easily profitable.
+    constexpr int64 kShapeThreshold = 128 * 2048;
+    constexpr int64 kInsnCountThreshold = 30;
+    auto root = insn.fused_expression_root();
+
+    // Too large shapes are not easily profitable.
+    if (root->opcode() == HloOpcode::kTuple) {
+      // Since all output shapes are the same, use the first shape as the
+      // representative.
+      auto shape = root->operand(0)->shape();
+      if (ShapeUtil::ElementsIn(shape) > kShapeThreshold) {
+        return false;
+      }
+    } else {
+      auto shape = root->shape();
+      if (ShapeUtil::ElementsIn(shape) > kShapeThreshold) {
+        return false;
+      }
+    }
+
+    // Having too many instructions is not easily profitable.
+    if (insn.fused_instruction_count() > kInsnCountThreshold) {
+      return false;
+    }
+
+    return true;
+  };
+
+  // Do not fuse computations with non-row-major layouts because it may
+  // increase bandwidth pressure.
+  auto all_row_major = [](const HloInstruction& fusion_insn) -> bool {
+    auto insns = fusion_insn.fused_instructions_computation()->instructions();
+    for (auto insn : insns) {
+      if (insn->shape().layout().format() != DENSE) {
+        continue;
+      }
+      if (!LayoutUtil::IsMonotonicWithDim0Major(insn->shape().layout())) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  for (auto insn : fusion_insns) {
+    if (!is_supported_fusion(*insn)) {
+      VLOG(2) << "Reject unsupported fusion insn " << insn->ToString();
+      continue;
+    } else if (!is_consumer_the_only_user(*insn)) {
+      VLOG(2) << "Reject maybe illegal insn " << insn->ToString();
+      continue;
+    } else if (!is_profitable(*insn)) {
+      VLOG(2) << "Reject may-not-be profitable fusion insn "
+              << insn->ToString();
+      continue;
+    } else if (!all_row_major(*insn)) {
+      VLOG(2) << "Reject non-row-major fusion insn " << insn->ToString();
+      continue;
+    } else {
+      fusion_insns_.push_back(insn);
+    }
+  }
+
+  // Sort according to the number of outputs and instruciton counts.
+  std::sort(fusion_insns_.begin(), fusion_insns_.end(),
+            [&](const HloInstruction* a, const HloInstruction* b) {
+              if (GetOutputSizeOfFusion(*a) == GetOutputSizeOfFusion(*b)) {
+                return a->fused_instruction_count() <
+                       b->fused_instruction_count();
+              }
+              return GetOutputSizeOfFusion(*a) < GetOutputSizeOfFusion(*b);
+            });
+}
+
+// Get a next span of fusion instructions to be fused.
+absl::Span<HloInstruction*>
+HorizontalFusionImpl::FusionCandidates::GetNextSpanOfFusions() {
+  if (pos_ >= fusion_insns_.size()) {
+    return absl::Span<HloInstruction*>();
+  }
+
+  // Fusing too many computations at a time may not be easily profitable and
```

Comment:
Did you see FusionWouldBeTooLarge? Would this be useful here too?
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/gpu/gpu_fusible.cc#L298

Reply:
I have considered it but there are some difficulties I see:
1. The horizontally fused kernel is much larger (e.g., in terms of the number of instructions) than the vertical fusions. So, they need different metrics.
2. We consider fusing many computations at a time, while `FusionWouldBeTooLarge` considers two fusion instructions at a time.